### ISSUE TITLE
Puppeteer, Jest, TypeError: Cannot read property 'add' of undefined

### ISSUE THREAD

**Issue by MatteoGioioso** [dfahlander/Dexie.js#790]
```
describe('IndexedDB', () => {
  jest.setTimeout(30000);
  const puppeteerConfig = {
    headless: false
  };
  let browser, page, db;

  beforeEach(async () => {
    browser = await puppeteer.launch(puppeteerConfig);
    page = await browser.newPage();
    await page.goto('http://localhost:3000', {waitUntil: 'networkidle2'});
    db = require('../../../services/cache/indexedDb')
  });

  afterEach(async () => {
    await db.clear();
    await browser.close();
  });


  describe('Respond Timeline photos', function () {
    const response = {
      Items: mocks.generatePhotosArray(9),
    };

    beforeEach(async() => {
      console.log(db)
      // await savePhotosCache(db, '/timeline', reducerNames.timelinePhotos,  response)

    });

    it('should respond with correct cache values', async function () {
      await db.photos.add({id: 1, date: '2014-01-09'})
    });
  });
});`
```
The db object is not empty 

```
 { default:
       Dexie {
         _dbSchema:
          { time: [TableSchema],
            photos: [TableSchema],
            albums: [TableSchema],
            albumsPhotos: [TableSchema],
            requests: [TableSchema],
            lastEvaluatedKey: [TableSchema] },
         version: [Function],
         _allTables:
          { time: [Table],
            photos: [Table],
            albums: [Table],
            albumsPhotos: [Table],
            requests: [Table],
            lastEvaluatedKey: [Table] },
         _createTransaction: [Function],
         _whenReady: [Function],
         verno: 1,
         open: [Function],
         close: [Function],
         delete: [Function],
         backendDB: [Function],
         isOpen: [Function],
         hasBeenClosed: [Function],
         hasFailed: [Function],
         dynamicallyOpened: [Function],
         name: 'test-store',
         on:
          { [Function: rv]
            addEventType: [Function: add],
            populate: [Object],
            blocked: [Object],
            versionchange: [Object],
            ready: [Object] },
         transaction: [Function],
         _transaction: [Function],
         table: [Function],
         Collection: [Function: Collection],
         Table: [Function: Table],
         Transaction: [Function: Transaction],
         Version: [Function: Version],
         WhereClause: [Function: WhereClause],
         time:
          Table {
            name: 'time',
            schema: [TableSchema],
            _tx: undefined,
            hook: [Function] },
         photos:
          Table {
            name: 'photos',
            schema: [TableSchema],
            _tx: undefined,
            hook: [Function] },
         albums:
          Table {
            name: 'albums',
            schema: [TableSchema],
            _tx: undefined,
            hook: [Function] },
         albumsPhotos:
          Table {
            name: 'albumsPhotos',
            schema: [TableSchema],
            _tx: undefined,
            hook: [Function] },
         requests:
          Table {
            name: 'requests',
            schema: [TableSchema],
            _tx: undefined,
            hook: [Function] },
         lastEvaluatedKey:
          Table {
            name: 'lastEvaluatedKey',
            schema: [TableSchema],
            _tx: undefined,
            hook: [Function] } } }
```

Everything is working normally, not with Puppeteer

**Comment by dfahlander** [comment]
That exception happens when there's no indexedDB property present on the global object, specifically i in node. Dexie can be configured to rely on a given indexedDB implementation instead, such as fake-indexedDB or indexeddbshim. To enable that, do set the static properties Dexie.dependencies.indexedDB and Dexie.dependencies.IDBKeyRange prior to instanciating your Dexie instance

**Comment by s0thl** [comment]
I'm facing the same problem right now. 

@dfahlander 

Could you please, post an example on how to maintain Dexie database instance from node? I have tried to initialize it from the renderer process first, then make use of it in the main process but still getting the `IndexedDB API not available` error.

**Comment by dfahlander** [comment]
Checkout this concluded solution: https://stackoverflow.com/questions/47934383/indexeddb-testing-with-jest-enzyme-referenceerror-indexeddb-is-not-defined

**Comment by s0thl** [comment]
Thanks, but I don't think this solution will work for me. I want to take the advantage of the embedded browser and use devtools to debug indexedDB, hence I have to pass real indexeddb. If you are familiar with the electron, do you think that using `global` for storing indexedDB will work well?

**Comment by dfahlander** [comment]
Ok if you are using Electron, you must use Dexie from within a BrowserWindow() instance only, or any web worker or service worker created from that BrowserWindow. You cannot use it from the main process. In BrowserWindow and the workers, indexedDB is available on their global contexts (window or self).

**Comment by s0thl** [comment]
You are right, but I'm talking about data access - like, there is often a need to communicate and exchange the data between several BrowserWindow instances. If each of window has it's own indexedDB, how should ones do that without using main as some kind of bridge? This is the only common process available out there. 

Consider this example:

We have following windows:

Main window that has two descendants:
   1. Account window,
   2. Game window

So, in total, we have three windows and the same amount of indexedDB's. In fact, we should be using only one indexedDB (that of the main window) to maintain everything, how do you overcome this, and, for instance, update account right from the `Game` BrowserWindow?

**Comment by dfahlander** [comment]
It's important to distinguish between a node process and a DOM thread. DOM threads can be windows or worker but not the main node process. Any DOM thread can have a database connection to the common database for all of them. Use a global Dexie instance on each DOM thread - preferrably declaring `const db = new Dexie(...); db.version(x).stores(...)` in a javascript module that you include from all of your windows and workers. But even if we polyfilled IndexedDB using IndexedDBShim or other, there are of my knowledge no possibility to access the same IndexedDB database from the node process.

Maybe you could consider to use a service worker that does central database related work instead of doing it from the node process? I'm not an expert on Electron, but I suppose there must be ways to communicate between the node process and a service worker some how...




 Each BrowserWindow will have their own Dexie instance, but it will be connected to the same database. The main process is not a BrowserWindow so it cannot access IndexedDB.

**Comment by dfahlander** [comment]
If you only want to work against the database from a certain worker, you can also use postMessage() to communicate against the worker and let it maintain the database communication. A library like [comlink](https://github.com/GoogleChromeLabs/comlink) to simplify that. Between windows and workers, use postMessage(). Between node main process and windows, use Electron's ipcMain. Don't know if there is a way to directly communicate directly between node process and a web- or service worker though, but you could always proxy that communication via a renderer.

**Comment by s0thl** [comment]
@dfahlander 

In electron, it's typically `main` and `renderer` process. Regarding communication, it's currently limited to the [IPC](https://electronjs.org/docs/api/ipc-main#sending-messages) commands, which makes it a pain to send the database data (indeed I would have to send request data from `main`, query & send back from `renderer` and finally receive on `main`, that is definitely not a way to go, especially when saving/updating the database records, that would require tons of commands). 

But, let's say I do base only on renderer processes and have a easy access to *own window* database, would it make it any easier to share the data between (only) renderer processes (that is, other windows)?

You said:

> Any DOM thread can have a database connection to the common database for all of them. Use a global Dexie instance on each DOM thread - preferrably declaring const db = new Dexie(...); db.version(x).stores(...)

Are you sure about that? I think each renderer process will establish a separated database and updating a database entry on renderer process one, wont update it on renderer process two.

**Comment by dfahlander** [comment]
I would be very surprised if different BrowserWindow instances would not share the same database. But I'm not an Electron expert. In a normal web scenario, the same database are shared across windows, iframes web workers and service workers as long as they are on the same origin (= hostname part of the URL).

**Comment by dfahlander** [comment]
If you create a `new Worker()` from within your browser window, or `navigator.serviceWorker.register()`, you would have workers that operates in the same origin, at least.

**Comment by s0thl** [comment]
> I would be very surprised if different BrowserWindow instances would not share the same database. But I'm not an Electron expert. In a normal web scenario, the same database are shared across windows, iframes web workers and service workers as long as they are on the same origin (= hostname part of the URL).

At least for me, they really don't. Thats why I am asking for any workaround, like accessing the Dexie singleton in *renderer one* from *renderer two*.

Also, I'm not using a worker yet. For now, I just open windows (one main + children) and try to find a way to read/write to database using Pixie from the main window's indexedDB (everything is being done on renderer process).
